apm-agent-plugins/apm-kafka-plugin/src/main/java/co/elastic/apm/agent/kafka/helper/ConsumerRecordsIteratorWrapper.java

```diff
@@ -0,0 +1,132 @@
+
+    public static final Logger logger = LoggerFactory.getLogger(ConsumerRecordsIteratorWrapper.class);
+
+    private final Iterator<ConsumerRecord> delegate;
+    private final ElasticApmTracer tracer;
+    private final CoreConfiguration coreConfiguration;
+    private final MessagingConfiguration messagingConfiguration;
+
+    public ConsumerRecordsIteratorWrapper(Iterator<ConsumerRecord> delegate, ElasticApmTracer tracer) {
+        this.delegate = delegate;
+        this.tracer = tracer;
+        coreConfiguration = tracer.getConfig(CoreConfiguration.class);
+        messagingConfiguration = tracer.getConfig(MessagingConfiguration.class);
+    }
+
+    @Override
+    public boolean hasNext() {
+        endCurrentTransaction();
+        return delegate.hasNext();
+    }
+
+    public void endCurrentTransaction() {
+        try {
+            Transaction transaction = tracer.currentTransaction();
+            if (transaction != null && "messaging".equals(transaction.getType())) {
+                transaction.deactivate().end();
+            }
+        } catch (Exception e) {
+            logger.error("Error in Kafka iterator wrapper", e);
+        }
+    }
+
+    @Override
+    public ConsumerRecord next() {
+        endCurrentTransaction();
+        ConsumerRecord record = delegate.next();
+        try {
+            String topic = record.topic();
+            if (!WildcardMatcher.isAnyMatch(messagingConfiguration.getIgnoreMessageQueues(), topic)) {
+                Header traceParentHeader = record.headers().lastHeader(TraceContext.TRACE_PARENT_HEADER);
+                Transaction transaction;
+                if (traceParentHeader != null) {
+                    transaction = tracer.startTransaction(
+                        TraceContext.fromTraceparentBinaryHeader(),
+                        traceParentHeader.value(),
+                        ConsumerRecordsIteratorWrapper.class.getClassLoader()
+                    );
+                } else {
+                    transaction = tracer.startRootTransaction(ConsumerRecordsIteratorWrapper.class.getClassLoader());
+                }
+                transaction.withType("messaging").withName("Kafka record from " + topic).activate();
+                Message message = transaction.getContext().getMessage();
+                message.withQueue(topic);
+                if (record.timestampType() == TimestampType.CREATE_TIME) {
+                    message.withAge(System.currentTimeMillis() - record.timestamp());
+                }
+                // todo - add destination fields
+
+                if (coreConfiguration.isCaptureHeaders()) {
+                    for (Header header : record.headers()) {
+                        String key = header.key();
+                        if (!TraceContext.TRACE_PARENT_HEADER.equals(key) &&
+                            WildcardMatcher.anyMatch(coreConfiguration.getSanitizeFieldNames(), key) == null) {
```

Comment:
Maybe sanitize centrally with a `co.elastic.apm.agent.report.processor.Processor`?

Reply:
Why? Especially given the comment above- better avoid this decoding if possible.

Reply:
So that we don't forget it on other messaging frameworks. What do you think about doing both maybe?

Reply:
I think it is a lot of overhead to pay in order to be future-proof, and requires special implementation for the new `Headers` type I am using. I rather not do that, at least at this point.